cypress/e2e/permit2.test.ts

```diff
@@ -0,0 +1,206 @@
+import { MaxUint160, MaxUint256 } from '@uniswap/permit2-sdk'
+
+import { DAI, USDC_MAINNET } from '../../src/constants/tokens'
+
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
+
+/** Initiates a swap and confirms its success. */
+const swaps = () => {
+  // Ensures that the swap button is enabled, since it can be temporarily disabled following approval.
+  cy.get('#swap-button').should('not.have.attr', 'disabled')
+
+  // Completes the swap.
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
+
+  // Verifies that there is a successful swap notification.
+  cy.contains('Swapped').should('exist')
+}
+
+// TODO: Update tests to differentiate between permit2 vs token approval button text once UI is updated to indicate approval step.
+describe('Permit2', () => {
+  // The same tokens & swap-amount combination is used for all permit2 tests.
+  const INPUT_TOKEN = DAI
+  const OUTPUT_TOKEN = USDC_MAINNET
+  const TEST_BALANCE_INCREMENT = 0.01
+
+  beforeEach(() => {
+    // Uses input/output in search params for faster test setup.
+    cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
+      ethereum: 'hardhat',
+    })
+    cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+  })
+
+  /** Asserts permit2 has a max approval for spend of the input token on-chain. */
+  const expectTokenAllowanceForPermit2ToBeMax = () => {
+    // check token approval
+    return cy
+      .hardhat()
+      .then(({ approval, wallet }) => approval.getTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN }))
+      .should('deep.equal', MaxUint256)
+  }
+
+  /** Asserts the universal router has a max permit2 approval for spend of the input token on-chain. */
+  const expectPermit2AllowanceForUniversalRouterToBeMax = (approvalTime: number) => {
+    return cy
+      .hardhat()
+      .then((hardhat) => hardhat.approval.getPermit2Allowance({ owner: hardhat.wallet, token: INPUT_TOKEN }))
+      .then((allowance) => {
+        cy.then(() => MaxUint160.eq(allowance.amount)).should('eq', true)
+        const expected = Math.floor((approvalTime + 2_592_000_000) / 1000)
+
+        // Asserts that the on-chain expiration is in 30 days, within a tolerance of 20 seconds.
+        cy.then(() => Math.abs(allowance.expiration - expected)).should('be.lessThan', 20)
```

Comment:
Use [`closeTo`](https://docs.cypress.io/guides/references/assertions#:~:text=closeTo(expected%2C%20delta)):

```suggestion
        cy.wrap(allowance.expiration).should('be.closeTo', expected, 20)
```